matview.c

```diff
@@ -1526,6 +1587,161 @@ rewrite_query_for_distinct_and_aggregates(Query *query, ParseState *pstate)
+							ereport(ERROR,
+									(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+									 errmsg("this query is not allowed on incrementally maintainable materialized view"),
+									 errhint("OR or NOT conditions and EXISTS condition are not used together")));
+						break;
+				}
+				break;
+			}
+		case T_SubLink:
+			{
+				char aliasName[NAMEDATALEN];
+				char columnName[NAMEDATALEN];
+				Query *subselect;
+				ParseState *pstate;
+				RangeTblEntry *rte;
+				RangeTblRef *rtr;
+				Alias *alias;
+				Oid opId;
+				ParseNamespaceItem *nsitem;
+
+				TargetEntry *tle_count;
+				FuncCall *fn;
+				Node *fn_node;
+				Expr *opexpr;
+
+				SubLink *sublink = (SubLink *)node;
+				/* raise ERROR if there is non-EXISTS sublink */
+				if (sublink->subLinkType != EXISTS_SUBLINK)
+					ereport(ERROR,
+							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							 errmsg("this query is not allowed on incrementally maintainable materialized view"),
+							 errhint("subquery in WHERE clause only supports subquery with EXISTS clause")));
+
+				subselect = (Query *)sublink->subselect;
+
+				/* raise ERROR if the sublink has CTE */
+				if (subselect->cteList)
+					ereport(ERROR,
+							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							 errmsg("CTE is not supported on incrementally maintainable materialized view")));
+
+				pstate = make_parsestate(NULL);
+				pstate->p_expr_kind = EXPR_KIND_SELECT_TARGET;
+
+				/*
+				 * convert EXISTS subquery into LATERAL subquery in FROM clause.
+				 */
+
+				snprintf(aliasName, sizeof(aliasName), "__ivm_exists_subquery_%d__", *count);
+				snprintf(columnName, sizeof(columnName), "__ivm_exists_count_%d__", *count);
+
+				/* add COUNT(*) for counting rows that meet exists condition */
+				fn = makeFuncCall(list_make1(makeString("count")), NIL, COERCE_EXPLICIT_CALL, -1);
+				fn->agg_star = true;
+				fn_node = ParseFuncOrColumn(pstate, fn->funcname, NIL, NULL, fn, false, -1);
+				tle_count = makeTargetEntry((Expr *) fn_node,
+											list_length(subselect->targetList) + 1,
+											columnName,
+											false);
+				/* add __ivm_exists_count__ column */
+				subselect->targetList = list_concat(subselect->targetList, list_make1(tle_count));
+				subselect->hasAggs = true;
+
+				/* add a sub-query whth LATERAL into from clause */
+				alias = makeAlias(aliasName, NIL);
+				nsitem = addRangeTableEntryForSubquery(pstate, subselect, alias, true, true);
+				rte = nsitem->p_rte;
+				query->rtable = lappend(query->rtable, rte);
+
+				/* assume the new RTE is at the end */
+				rtr = makeNode(RangeTblRef);
+				rtr->rtindex = list_length(query->rtable);
+				((FromExpr *)query->jointree)->fromlist = lappend(((FromExpr *)query->jointree)->fromlist, rtr);
+
+				/*
+				 * EXISTS condition is converted to HAVING count(*) > 0.
+				 * We use make_opcllause() to get int84gt( '>' operator). We might be able to use make_op().
+				 */
+				opId = OpernameGetOprid(list_make1(makeString(">")), INT8OID, INT4OID);
```

Comment:
This operator must be qualified, too. Or, maybe, there is more clever way to get the desired operator id ... ? (I have no idea for now, though.)

Reply:
Fixed to process  like SystemFuncName().